src/AbstractSessionGenerator.php

```diff
@@ -64,54 +60,59 @@ protected function _generateRecursive(
+        $startTimes = [$start];
+        $results    = [];
+        $processed  = [];
 
-            // If exceeded range end, no more iteration is needed.
-            if ($_sessionEnd > $end) {
-                break;
-            }
+        GENERATE_SESSIONS:
 
-            // Create session using the factory
-            $session = call_user_func_array($factory, [$start, $_sessionEnd]);
+        $newStartTimes = [];
+        foreach ($startTimes as $_start) {
+            foreach ($lengths as $_length) {
+                // Calculate end of session from range start time
+                $_sessionEnd = $_start + $_length;
 
-            try {
-                // Validate session and add to results
-                if ($validator !== null) {
-                    $validator->validate($session);
+                // If exceeded range end, no more iteration is needed.
+                if ($_sessionEnd > $end) {
+                    break;
                 }
-                $results[]         = $session;
-                $processed[$start] = true;
-            } catch (ValidationFailedExceptionInterface $exception) {
-                if ($invalidCb !== null) {
-                    call_user_func_array($invalidCb, [$session, $exception]);
+
+                // Create session using the factory
+                $session = call_user_func_array($factory, [$_start, $_sessionEnd]);
+
+                try {
+                    // Validate session and add to results
+                    if ($validator !== null) {
+                        $validator->validate($session);
+                    }
+                    $results[]          = $session;
+                    $processed[$_start] = true;
+                } catch (ValidationFailedExceptionInterface $exception) {
+                    if ($invalidCb !== null) {
+                        call_user_func_array($invalidCb, [$session, $exception]);
+                    }
                 }
-            }
 
-            // Calculate the next start time for recursion, skipping if already generated
-            $_next = $_sessionEnd + $padding;
+                // Calculate the next start time for recursion, skipping if already generated
+                $_next = $_sessionEnd + $padding;
 
-            if (isset($processed[$_next])) {
-                continue;
+                if (isset($processed[$_next])) {
+                    continue;
+                }
+
+                $newStartTimes[] = $_next;
             }
+        }
 
-            $this->_generateRecursive(
-                $_next,
-                $end,
-                $lengths,
-                $padding,
-                $factory,
-                $validator,
-                $invalidCb,
-                $results,
-                $processed
-            );
+        $startTimes = $newStartTimes;
+
+        if (count($newStartTimes) === 0) {
+            return $results;
         }
+
+        goto GENERATE_SESSIONS;
```

Comment:
Could this be done using a loop instead?

Reply:
Perhaps. Didn't try tbh. Since this is an approach that results in an optimization that goes all the way down to op-code, I'm a bit reluctant to try it.

Reply:
Ok, maybe l8r.